### PR TITLE
feat: parse markdown headings into a nested Section graph

### PR DESCRIPTION
**Pull request by vitali87**
## Summary

Markdown files were indexed as a bare `File` node, so headings — the structure people actually organise plan and spec files around — were invisible to the graph.

This adds a **document tier** alongside the tree-sitter and ast-grep tiers. Each heading becomes a `Section` node carrying its text, heading level (1-6) and line span, nested through a new `CONTAINS_SECTION` edge so a subheading hangs off the heading above it and a top-level heading hangs off the file's `Module`.

Indexing a mixed document produces:

```
Sections emitted: 5
  L  1  h1  'Architecture'      qn=proj.docs.arch.Architecture
  L  5  h2  'Ingestion'         qn=proj.docs.arch.Architecture.Ingestion
  L  9  h3  'Parsers'           qn=proj.docs.arch.Architecture.Ingestion.Parsers
  L 13  h2  'Querying'          qn=proj.docs.arch.Architecture.Querying
  L 17  h1  'Appendix'          qn=proj.docs.arch.Appendix        <- setext, back to module level
```

## Two design decisions worth reviewing

**Nesting comes from heading levels, not the grammar's `section` nodes.** I probed the grammar rather than assuming, and ATX and setext headings parse differently:

```
section                        # '## Phase One' NESTS as a section node
  atx_heading
  section
    atx_heading                # '### Subtask A' nests one deeper

section                        # setext is FLAT:
  setext_heading  'Title'      #   h1 and h2 are SIBLINGS,
  paragraph                    #   and the text sits one level deeper,
  setext_heading  'Sub'        #   inside a paragraph, not a direct inline child
```

Level arithmetic is the only rule that handles both forms alike. It also gives skipped levels (`#` straight to `###`) the natural answer: the deeper heading is a child of whatever heading is currently open. Both setext traps are silent — a naive direct-children reader returns empty text and drops every setext heading — so they are pinned by tests.

**Documents get their own tier rather than reusing an existing one.** `LanguageSpec` is built around `function_node_types` / `class_node_types` / `call_node_types`; mapping a heading onto any of those would make every heading a `Class` in the graph and surface headings in `cgr dead-code` and `cgr duplicates` output. The ast-grep tier emits a flat Module/Function/Class set with no nesting, which would discard the parent/child structure that is the entire point. The precedent for a tier that call-graph consumers skip is already there in `structural_tier_extensions()`.

## Two registries that are easy to miss

Adding a `NodeLabel` is not a one-line change, and an existing invariant test caught the first of these:

- **`NODE_SCHEMAS` / `RELATIONSHIP_SCHEMAS`** (`types_defs.py`) feed the schema the agent sees, README generation, and `graph_audit`. `test_every_documented_label_has_parsed_properties` fails loudly if a label is missing — good guard.
- **`CYPHER_DELETE_PROJECT`** walks containment to delete a project's nodes. Without `CONTAINS_SECTION` in that traversal, `cgr delete-project` would leave every `Section` orphaned in the graph. Nothing tests this at unit level, so it is worth a look.

## Degraded install

The grammar ships in the `treesitter-full` extra. `handles()` returns False when it is absent, so a base install falls through to the generic `File` node rather than failing — verified against an environment with the grammar uninstalled, and covered by tests.

## Test Plan

- New `codebase_rag/tests/test_document_tier.py`, 25 tests, indexing real `.md` files end to end through `GraphUpdater`: heading extraction, levels, 1-based line spans, nesting, sibling-closes-deeper, skipped levels, all four setext cases, qualified-name collisions, `File`/`Module` coexistence, `.markdown`, nested directories, and the three degraded-grammar paths.
- Full unit suite: **7885 passed, 120 skipped**. `ruff check`, `ruff format --check` and `ty check` clean on the new files.
- `scripts/generate_readme.py` is a no-op against the hand-written doc sections.

Edge cases pinned by tests, each of which fails silently otherwise:

| Case | Behaviour |
|---|---|
| Prose before any heading | Grammar wraps it in a headingless `section`; emitting it would put a nameless node in the graph, so it is skipped |
| Two `## Notes` under one parent | Same qn would merge two sections into one node; disambiguated with the existing `<qn>@<line>` convention |
| `# Release 1.2.3` | Dots would create phantom hierarchy levels when a qn is split; sanitised in the qn, kept in the display name |
| Document with no headings | No Section nodes, `File` node unaffected |

## Related Issues

Implements #1426. Splits the structural half of #164 — the behavioural half (dedup, plan consolidation, purpose awareness) stays on that issue.

Found while working on this and filed separately rather than fixed here: **#1427** — the realtime watcher re-parses only tree-sitter files, so ast-grep-tier languages already go stale on live edits and this tier inherits that. It predates this PR; the right fix is factoring out the shared tier dispatch, which is more than this change should carry.

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] Test updates
- [x] Documentation update

## Summary by CodeRabbit

- **New Features**
  - Added Markdown document indexing with nested sections.
  - Captures heading text, levels, and source line ranges through each section’s content.
  - Supports ATX and Setext headings, duplicate names, and Markdown variants.
  - Preserves document and file records when enhanced parsing is unavailable.
  - Added optional Markdown parsing support.
  - Sections are represented in exported graph data with containment relationships.
  - Removes indexed sections when their source document is deleted.

- **Documentation**
  - Documented section structure, nesting, Markdown support, and call-graph limitations.

### CODE REVIEW

**Comment by vitali87** [comment]
@greptileai review

File: codebase_rag/parsers/document_tier.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+                cs.KEY_PATH: relative_path,
+                cs.KEY_ABSOLUTE_PATH: absolute_path,
+            },
+        )
+        parent_label = cs.NodeLabel.MODULE if parent_is_module else cs.NodeLabel.SECTION
+        self._ingestor.ensure_relationship_batch(
+            (parent_label, cs.KEY_QUALIFIED_NAME, parent_qn),
+            cs.RelationshipType.CONTAINS_SECTION,
+            (cs.NodeLabel.SECTION, cs.KEY_QUALIFIED_NAME, qualified_name),
+        )
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/document_tier.py:304]: Correct and fixed. CYPHER_DELETE_MODULE walked only DEFINES|DEFINES_METHOD, so the delete-before-reingest path dropped the Module and stranded every Section under it — stale headings would accumulate on each re-index of an edited document. CONTAINS_SECTION is now in that walk. Same class of bug as the CYPHER_DELETE_PROJECT gap already fixed in this PR; thanks for catching the per-file one.

Note the live-watch case you allude to is a separate, pre-existing gap: realtime_updater.py re-parses only tree-sitter files, so ast-grep-tier languages (Ruby, Kotlin, ...) already go stale on edit today. Filed as #1427 rather than fixed here, since the right fix is factoring out the shared tier dispatch.

File: codebase_rag/constants/graph.py [hunk 2]
```diff
@@ -120,6 +120,9 @@ class NodeLabel(StrEnum):
     EXTERNAL_PACKAGE = "ExternalPackage"
     EXTERNAL_MODULE = "ExternalModule"
     RESOURCE = "Resource"
+    # A heading and the prose beneath it in a document (issue #1426). Sections
+    # nest by heading level, so a Section CONTAINS_SECTION its subsections.
+    SECTION = "Section"
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/constants/graph.py:126]: Correct and fixed. Verified the mechanism first: ProtobufFileIngestor.ensure_node_batch drops any label with no LABEL_TO_ONEOF_FIELD entry (warning only, node discarded), so `cgr index --output` on a repo with Markdown would have silently lost every Section.

Added a Section message plus `Section section = 18` in the Node oneof, CONTAINS_SECTION = 18 in the RelationshipType enum, and the NodeLabel.SECTION -> ONEOF_SECTION mapping. Both use next-available field numbers, so the change is backward compatible. Bindings regenerated with protobuf 6.33.5 to match the pinned runtime; I checked the toolchain reproduces the previously checked-in schema_pb2.py byte-for-byte before regenerating, so the diff is only the additions.

Covered by three round-trip tests that index a real document, serialise, deserialise, and assert the Section payloads and typed edges survive. I confirmed they fail without the mapping rather than passing vacuously.

**Comment by vitali87** [comment]
@greptileai review

File: codebase_rag/parsers/document_tier.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+            if qualified_name in claimed_qns:
+                qualified_name = f"{qualified_name}{cs.DUP_QN_MARKER}{start_line}"
+            claimed_qns.add(qualified_name)
+
+            self._emit_section(
+                qualified_name=qualified_name,
+                name=text or _UNTITLED,
+                level=level,
+                start_line=start_line,
+                end_line=heading.end_point[0] + 1,
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/document_tier.py]: Correct, and this one was a real defect — the module docstring promised "the line span of the heading and the prose beneath it" while the code stored only the heading node. Verified before fixing: a 10-line document reported Top as 1-2 and Child as 7-8.

Spans are now computed by lookahead over the pre-collected headings: a section ends at the line before the next heading at the same or a shallower level, or at EOF. Same document now gives Top 1-10 and Child 7-10.

One deliberate departure from the suggestion: a DEEPER heading does not close a section, so a parent span contains its subsections (Top 1-9 covers its two h2 children). That matches the containment CONTAINS_SECTION already describes, and treating a child as a terminator would make the parent span exclude its own subtree. Four tests pin it — the prose case, sibling closing, parent-contains-children, and the EOF case where a trailing newline must not invent a line. All four fail against the old heading-node span.

File: codebase_rag/parsers/document_tier.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+
+            while open_headings and open_headings[-1][0] >= level:
+                open_headings.pop()
+
+            parent_qn = open_headings[-1][1] if open_headings else module_qn
+            start_line = heading.start_point[0] + 1
+            qualified_name = f"{parent_qn}{cs.SEPARATOR_DOT}{name}"
+            if qualified_name in claimed_qns:
+                qualified_name = f"{qualified_name}{cs.DUP_QN_MARKER}{start_line}"
+            claimed_qns.add(qualified_name)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/document_tier.py:244]: Right, and adopted — the single suffix pass is now a while loop.

Worth recording the actual reproduction, because the scenario in the comment does not collide: with a literal "## Notes@5" heading present, a later duplicate "Notes" on line 5 claims the UNSUFFIXED name first, so "Notes@5" is never generated for it. The collision needs the plain name claimed first, then the literal marker heading, then the duplicate: "## Notes", "## Notes@9", then "## Notes" starting on line 9. That third heading generates the already-owned "Notes@9" and the two sections merge into one node.

My first attempt at a regression test used the scenario as described and passed with and without the fix, so I rebuilt it around the reproduction above; it now fails with `if` and passes with `while`.

File: codebase_rag/tests/test_document_tier.py [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+
+        index = pb.GraphCodeIndex()
+        index.ParseFromString((out / "index.bin").read_bytes())
+        section_rels = [r for r in index.relationships if r.target_label == SECTION]
+        assert section_rels, "no relationships target a Section"
+        unspecified = pb.Relationship.RelationshipType.RELATIONSHIP_TYPE_UNSPECIFIED
+        for rel in section_rels:
+            assert rel.type != unspecified, (
+                f"CONTAINS_SECTION serialised as UNSPECIFIED: {rel}"
+            )
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/tests/test_document_tier.py:404]: Fair — the not-UNSPECIFIED check was weaker than it should be, since any mapped enum value would have satisfied it. Now asserts equality with pb.Relationship.RelationshipType.CONTAINS_SECTION.

**Comment by vitali87** [comment]
@greptileai review

File: codebase_rag/parsers/document_tier.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+    A section runs until the next heading at the same or a shallower level
+    (a deeper heading is its child and stays inside its span), or to the end
+    of the file when nothing closes it. Using the heading node's own end
+    instead would report a one- or two-line span for every section.
+    """
+    for next_heading, next_level in levelled[index + 1 :]:
+        if next_level <= level:
+            # The line before the closing heading; a heading immediately
+            # after another leaves the parent owning only its own line.
+            return max(next_heading.start_point[0], levelled[index][0].end_point[0] + 1)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/document_tier.py]: Good catch, adopted. The slice copied the tail for every heading, so the copying was quadratic across a heading-dense document even though the loop almost always stops on the very first entry. Now indexed via range().

Also hoisted `levelled[index][0].end_point[0] + 1` into a local: both return paths recomputed it identically.

Verified behaviour is unchanged rather than assuming it: a 200-heading document still produces correct spans (H0 closes at line 8 exactly before H1 opens at 9, with its nested S0 inside), and all 33 tier tests pass.

**Comment by vitali87** [comment]
@greptileai review

**Comment by vitali87** [comment]
@greptileai review

File: codebase_rag/parsers/flat_module.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    """Emit a file's Module node and its containment edge; return its qn.
+
+    The qualified name is flat: no ``__init__``/``mod`` special case and no
+    stem disambiguation, so two files whose stems collide in one directory
+    would share a qn. Add disambiguation here if a tier ever needs it.
+    """
+    relative_path = cached_relative_path(file_path, repo_path)
+    module_qn = cs.SEPARATOR_DOT.join(
+        [project_name, *relative_path.with_suffix("").parts]
+    )
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/flat_module.py]: Verified and fixed for this tier. Confirmed the collision first: docs/guide.md and docs/guide.markdown both produced qn <proj>.docs.guide, so the two Modules merged and their identically-named Sections merged with them.

The document tier now passes distinguish_suffix=True, giving guide_md and guide_markdown (underscore, not a dot — a dot would read as another hierarchy level to every consumer that splits a qn). Two files with identical headings now yield 2 distinct Modules and 2 distinct Sections, covered by a test.

Deliberately scoped narrower than the comment suggests, though. This is NOT new: the ast-grep tier has always had it, and build.sh/build.bash merge the same way today — its own comment even said "add if a config language collides with another file stem in the same directory". Kotlin (.kt/.kts) and Elixir (.ex/.exs) are the realistic pairs.

So the flag is opt-in rather than the default. Turning it on for ast-grep changes the qn of EVERY module across its eight languages (app.rb -> app_rb, not just colliding pairs), which affects prefix-matching Cypher queries, import resolution, and requires a re-index for existing graphs. That is a cross-language behaviour change that deserves its own review rather than riding inside a Markdown feature, so I filed it as #1429 with the reproduction and a checklist. Confirmed ast-grep qns are byte-identical after this change (app.rb still app).

**Comment by vitali87** [comment]
@greptileai review

**Comment by vitali87** [comment]
@greptileai review
